src/pybind/mgr/cephadm/module.py

```diff
@@ -324,10 +333,194 @@ def __init__(self, *args, **kwargs):
+                        'value': target_name,
+                        'who': daemon_type + '.' + d.service_instance,
+                    })
+                    return self._service_action([(
+                        d.service_type,
+                        d.service_instance,
+                        d.nodename,
+                        'redeploy'
+                    )])
+
+            if need_upgrade_self:
+                mgr_map = self.get('mgr_map')
+                num = len(mgr_map.get('standbys'))
+                if not num:
+                    self.log.warning(
+                        'Upgrade: No standby mgrs and I need to update the mgr, '
+                        'suspending upgrade')
+                    self.upgrade_state['error'] = 'No standby mgrs and mgr.%s ' \
+                        'needs to be upgraded' % self.get_mgr_id()
+                    return None
+
+                self.log.info('Upgrade: there are %d other already-upgraded '
+                              'standby mgrs, failing over' % num)
+
+                # fail over
+                ret, out, err = self.mon_command({
+                    'prefix': 'mgr fail',
+                    'who': self.get_mgr_id(),
+                })
+                return None
+
+            # make sure 'ceph versions' agrees
+            ret, out, err = self.mon_command({
+                'prefix': 'versions',
+            })
+            j = json.loads(out)
+            self.log.debug('j %s' % j)
+            for version, count in j.get(daemon_type, {}).items():
+                if version != target_version:
+                    self.log.warning(
+                        'Upgrade: %d %s daemon(s) are %s != target %s' %
+                        (count, daemon_type, version, target_version))
+
+            # push down configs
+            if image_settings.get(daemon_type) != target_name:
+                self.log.info('Upgrade: Setting container_image for all %s...' %
+                              daemon_type)
+                ret, out, err = self.mon_command({
+                    'prefix': 'config set',
+                    'name': 'container_image',
+                    'value': target_name,
+                    'who': daemon_type,
+                })
+            to_clean = []
+            for section in image_settings.keys():
+                if section.startswith(daemon_type + '.'):
+                    to_clean.append(section)
+            if to_clean:
+                self.log.info('Upgrade: Cleaning up container_image for %s...' %
+                              to_clean)
+                for section in to_clean:
+                    ret, image, err = self.mon_command({
+                        'prefix': 'config rm',
+                        'name': 'container_image',
+                        'who': section,
+                    })
+            self.log.info('Upgrade: All %s daemons are up to date.' %
+                          daemon_type)
```

Comment:
If we rely on the default image being set correctly, maybe we can remove this code block?